aptos-move/dynamic-transaction-composer/src/codegen.rs

```diff
@@ -0,0 +1,380 @@
+}
+
+impl Context {
+    // Add signers to the beginning of the script's parameter pool
+    fn add_signers(&mut self, signer_counts: u16) -> PartialVMResult<()> {
+        for _ in 0..signer_counts {
+            self.parameters
+                .push(SignatureToken::Reference(Box::new(SignatureToken::Signer)));
+        }
+        self.signer_counts = signer_counts;
+        Ok(())
+    }
+
+    // Iterate through the batched calls and convert all arguments in the BatchedFunctionCall into
+    // arguments to the script.
+    fn allocate_parameters(&mut self, calls: &[BatchedFunctionCall]) -> PartialVMResult<()> {
+        let mut total_args_count = self.parameters.len() as u16;
+        for call in calls {
+            self.args_to_local.push(total_args_count);
+            for arg in call.args.iter() {
+                // Only the serialized arguments need to be allocated as input to the script.
+                // Return values from previous calls will be stored in the locals.
+                if let BatchArgument::Raw(bytes) = arg {
+                    if total_args_count == TableIndex::MAX {
+                        return Err(PartialVMError::new(StatusCode::INDEX_OUT_OF_BOUNDS));
+                    }
+                    self.args.push(bytes.clone());
+                    total_args_count += 1;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    // Generate instructions for invoking one batched call.
+    fn compile_batched_call(
+        &mut self,
+        current_call_idx: usize,
+        call: &BatchedFunctionCall,
+        func_id: FunctionHandleIndex,
+    ) -> PartialVMResult<()> {
+        let func_handle = self.script.function_handle_at(func_id).clone();
+
+        // Allocate generic type parameters to the script
+        //
+        // Creates mapping from local type argument index to type argument index in the script
+        let mut subst_mapping = BTreeMap::new();
+        for (idx, ty_param) in self.ty_args[current_call_idx].iter().enumerate() {
+            subst_mapping.insert(idx as u16, ty_param.clone());
+        }
+
+        // Instructions for loading parameters
+        for (idx, arg) in call.args.iter().enumerate() {
+            match arg {
+                BatchArgument::PreviousResult(PreviousResult {
+                    call_idx,
+                    return_idx,
+                    operation_type,
+                }) => {
+                    if let Some(idx) = self.returned_val_to_local.get(*call_idx as usize) {
+                        if *return_idx >= self.return_counts[*call_idx as usize] {
+                            return Err(PartialVMError::new(StatusCode::INDEX_OUT_OF_BOUNDS));
+                        }
+                        // Locals will need to be offset by:
+                        // 1. The parameters to the script (signer + arguments)
+                        // 2. All the previously returned values prior to the call.
+                        let local_idx =
```

Comment:
I'm wondering whether it would be perhaps easier to reason and extend over time that, instead of this indexing, you first make a walk over the list of calls and assign locals. You basically have another variant in `Argument`, like `ResolvedLocal(index)`. This for now refers to the previous result.

Over time, we can have more complex dependencies and the 2nd phase, actually generating bytecode, stays unchanged. Also think it could be already now simpler.

Reply:
Actually the current code works like what you are suggesting: we first walk over the list of calls and assign locals for each of the individual arguments. The code how it was organized is a bit messy tho, in the current impl we first scan the list for locals and then during codegen for each calls we push the arguments to the local list. So in the scripts, the arguments are indeed encoded as arguments to the scripts, instead of inlining them. 

Another interesting side note is I did actually inlined the type arguments. So the generated script will have no generic type parameters. This kinda acts like monomorphization.

Reply:
I think it would be cleaner to assign the locals during allocation:

```
for arg in call_args.iter_mut() {
   ...
   *arg = Argument::Resolved(..)
}
```

Then you can get rid of all those ugly indexing, plus the subsequent codegen can express arbitrary relations not only the current fixed ones. 

One thing is you can just leverage Rust mutation as above, ugly thing here is you add a variant to `Argument` which is public which is in fact internal. What conceptually happens that you have a little AST and annotate each node with additional info, like the locals assign to each argument and result. So another approach is to give each call a `CallId(usize)`. Then have a `Map<CallId, CallInfo>` with `struct CallInfo { inputs: Vec<LocalIndex>, outputs: Vec<LocalIndex> }`. This is then input to the actual codegen.

I believe this leads as easily to a more general representation which supports control flow and essentially arbitrary code.